File: android/src/main/kotlin/com/anisrehman/archarts/LineChartPlatformView.kt
package com.anisrehman.archarts

import android.content.Context
import android.view.View
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.highlight.Highlight
import io.flutter.plugin.platform.PlatformView

class LineChartPlatformView(
    context: Context,
    private val params: Map<String, Any?>
) : PlatformView {

    private val chart: LineChart = LineChart(context)

    init {
        applyConfig()
    }

    override fun getView(): View = chart

    override fun dispose() = Unit

    private fun applyConfig() {
        val series = params["series"] as? List<*> ?: emptyList<Any>()
        val defaultStyle = params["defaultLineStyle"] as? Map<String, Any?>
        val perSeriesStyle = params["perSeriesStyle"] as? Map<String, Any?>

        val dataSets = series.mapNotNull { item ->
            val seriesMap = item as? Map<String, Any?> ?: return@mapNotNull null
            val points = seriesMap["points"] as? List<*> ?: emptyList<Any>()
            val entries = points.mapNotNull { point ->
                val pointMap = point as? Map<String, Any?> ?: return@mapNotNull null
                val x = (pointMap["x"] as? Number)?.toFloat() ?: return@mapNotNull null
                val y = (pointMap["y"] as? Number)?.toFloat() ?: return@mapNotNull null
                Entry(x, y)
            }
            val label = seriesMap["label"] as? String
            val dataSet = LineDataSet(entries, label ?: "")
            val seriesId = seriesMap["id"] as? String
            val styleMap = seriesId?.let { id -> perSeriesStyle?.get(id) as? Map<String, Any?> }
                ?: defaultStyle
            applyLineStyle(dataSet, styleMap)
            dataSet
        }

        chart.data = LineData(dataSets)

        applyAxis(chart.xAxis, params["xAxis"] as? Map<String, Any?>, AxisType.X)
        applyAxis(chart.axisLeft, params["leftAxis"] as? Map<String, Any?>, AxisType.Y)
        applyAxis(chart.axisRight, params["rightAxis"] as? Map<String, Any?>, AxisType.Y)

        applyLegend(params["legend"] as? Map<String, Any?>)
        applyInteraction(params["interaction"] as? Map<String, Any?>)
        applyViewport(params["viewport"] as? Map<String, Any?>)
        applyMarker(params["marker"] as? Map<String, Any?>)
        applyAnimation(params["animation"] as? Map<String, Any?>)

        chart.description.isEnabled = false
        chart.invalidate()
    }

    // Keep in sync with iOS LineChartPlatformView.applyLineStyle (same order and rules).
    private fun applyLineStyle(dataSet: LineDataSet, styleMap: Map<String, Any?>?) {
        if (styleMap == null) return

        val lineColor = (styleMap["lineColor"] as? Number)?.toInt()
        if (lineColor != null) {
            dataSet.color = lineColor
        }
        val lineWidth = (styleMap["lineWidth"] as? Number)?.toFloat()
        if (lineWidth != null) {
            dataSet.lineWidth = lineWidth
        }
        val drawCircles = styleMap["drawCircles"] as? Boolean
        if (drawCircles != null) {
            dataSet.setDrawCircles(drawCircles)
        }
        val circleColor = (styleMap["circleColor"] as? Number)?.toInt()
        if (circleColor != null) {
            dataSet.setCircleColor(circleColor)
        } else if (lineColor != null) {
            dataSet.setCircleColor(lineColor)
        }
        val circleRadius = (styleMap["circleRadius"] as? Number)?.toFloat()
        if (circleRadius != null) {
            dataSet.circleRadius = circleRadius
            // radius <= 0 means no visible circles, so disable drawing.
            if (circleRadius <= 0f) dataSet.setDrawCircles(false)
        }
        val drawValues = styleMap["drawValues"] as? Boolean
        if (drawValues != null) {
            dataSet.setDrawValues(drawValues)
        }
        val cubic = styleMap["cubic"] as? Boolean
        if (cubic == true) {
            dataSet.mode = LineDataSet.Mode.CUBIC_BEZIER
        }
    }

    private fun applyAxis(axis: XAxis, axisMap: Map<String, Any?>?, type: AxisType) {
        if (axisMap == null) return
        axis.isEnabled = axisMap["enabled"] as? Boolean ?: true
        axis.setDrawGridLines(axisMap["drawGridLines"] as? Boolean ?: true)
        axis.setDrawAxisLine(axisMap["drawAxisLine"] as? Boolean ?: true)

        val min = (axisMap["min"] as? Number)?.toFloat()
        val max = (axisMap["max"] as? Number)?.toFloat()
        if (min != null) axis.axisMinimum = min
        if (max != null) axis.axisMaximum = max

        val labelCount = axisMap["labelCount"] as? Int
        if (labelCount != null) {
            axis.setLabelCount(labelCount, true)
        }

        if (type == AxisType.X) {
            axis.position = XAxis.XAxisPosition.BOTTOM
        }
    }

    private fun applyAxis(axis: YAxis, axisMap: Map<String, Any?>?, type: AxisType) {
        if (axisMap == null) return
        axis.isEnabled = axisMap["enabled"] as? Boolean ?: true
        axis.setDrawGridLines(axisMap["drawGridLines"] as? Boolean ?: true)
        axis.setDrawAxisLine(axisMap["drawAxisLine"] as? Boolean ?: true)

        val min = (axisMap["min"] as? Number)?.toFloat()
        val max = (axisMap["max"] as? Number)?.toFloat()
        if (min != null) axis.axisMinimum = min
        if (max != null) axis.axisMaximum = max

        val labelCount = axisMap["labelCount"] as? Int
        if (labelCount != null) {
            axis.setLabelCount(labelCount, true)
        }

        YAxisValueFormatter.fromAxisMap(axisMap)?.let { formatter ->
            axis.valueFormatter = formatter
        }
    }

    private fun applyLegend(legendMap: Map<String, Any?>?) {
        val legend = chart.legend
        if (legendMap == null) return
        legend.isEnabled = legendMap["enabled"] as? Boolean ?: true

        when (legendMap["position"] as? String) {
            "bottom" -> legend.verticalAlignment = Legend.LegendVerticalAlignment.BOTTOM
            "left" -> legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
            "right" -> legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
            else -> legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
        }

        when (legendMap["position"] as? String) {
            "left" -> legend.horizontalAlignment = Legend.LegendHorizontalAlignment.LEFT
            "right" -> legend.horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
            else -> legend.horizontalAlignment = Legend.LegendHorizontalAlignment.CENTER
        }

        legend.orientation = Legend.LegendOrientation.HORIZONTAL
        legend.setDrawInside(false)
    }

    private fun applyInteraction(interactionMap: Map<String, Any?>?) {
        if (interactionMap == null) return
        val zoomEnabled = interactionMap["zoomEnabled"] as? Boolean ?: true
        val dragEnabled = interactionMap["dragEnabled"] as? Boolean ?: true
        val highlightEnabled = interactionMap["highlightEnabled"] as? Boolean ?: true

        chart.setScaleEnabled(zoomEnabled)
        chart.setPinchZoom(zoomEnabled)
        chart.isDragEnabled = dragEnabled
        chart.isHighlightPerTapEnabled = highlightEnabled
        chart.isHighlightPerDragEnabled = highlightEnabled
    }

    private fun applyViewport(viewportMap: Map<String, Any?>?) {
        if (viewportMap == null) return
        val minRange = (viewportMap["visibleXRangeMin"] as? Number)?.toFloat()
        if (minRange != null) {
            chart.setVisibleXRangeMinimum(minRange)
        }
        val maxRange = (viewportMap["visibleXRangeMax"] as? Number)?.toFloat()
        if (maxRange != null) {
            chart.setVisibleXRangeMaximum(maxRange)
        }
        val initialX = (viewportMap["initialX"] as? Number)?.toFloat()
        if (initialX != null) {
            chart.moveViewToX(initialX)
        }
        val offsets = viewportMap["viewPortOffsets"] as? Map<*, *>
        if (offsets != null) {
            val left = (offsets["left"] as? Number)?.toFloat() ?: 0f
            val top = (offsets["top"] as? Number)?.toFloat() ?: 0f
            val right = (offsets["right"] as? Number)?.toFloat() ?: 0f
            val bottom = (offsets["bottom"] as? Number)?.toFloat() ?: 0f
            chart.setViewPortOffsets(left, top, right, bottom)
        }
    }

    private fun applyMarker(markerMap: Map<String, Any?>?) {
        if (markerMap == null) return
        val enabled = markerMap["enabled"] as? Boolean ?: false
        if (!enabled) return
        val format = markerMap["format"] as? String
        chart.marker = ChartMarkerView(chart.context, format)
    }

    private fun applyAnimation(animationMap: Map<String, Any?>?) {
        if (animationMap == null) return
        val enabled = animationMap["enabled"] as? Boolean ?: false
        if (!enabled) return
        val duration = (animationMap["durationMs"] as? Number)?.toInt() ?: 500
        when (animationMap["easing"] as? String) {
            "linear" -> chart.animateX(duration, com.github.mikephil.charting.animation.Easing.Linear)
            else -> chart.animateX(duration, com.github.mikephil.charting.animation.Easing.EaseInOutQuad)
        }
    }

    private enum class AxisType { X, Y }
}

File: ios/ar_charts/Sources/ar_charts/LineChartPlatformView.swift
import Flutter
import UIKit
import DGCharts

final class LineChartViewFactory: NSObject, FlutterPlatformViewFactory {
    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        let params = args as? [String: Any] ?? [:]
        return LineChartPlatformView(frame: frame, viewId: viewId, params: params)
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

final class LineChartPlatformView: NSObject, FlutterPlatformView {
    private let chartView: LineChartView
    private let viewId: Int64

    init(frame: CGRect, viewId: Int64, params: [String: Any]) {
        self.chartView = LineChartView(frame: frame)
        self.viewId = viewId
        super.init()
        ChartViewRegistry.registerLineChart(viewId: viewId, view: self)
        applyConfig(params: params)
    }

    deinit {
        ChartViewRegistry.unregisterLineChart(viewId: viewId)
    }

    func view() -> UIView {
        return chartView
    }

    func updateConfig(params: [String: Any]) {
        applyConfig(params: params)
    }

    private func applyConfig(params: [String: Any]) {
        let series = params["series"] as? [[String: Any]] ?? []
        let defaultStyle = params["defaultLineStyle"] as? [String: Any]
        let perSeriesStyle = params["perSeriesStyle"] as? [String: Any]

        var dataSets: [LineChartDataSet] = []
        for item in series {
            let points = item["points"] as? [[String: Any]] ?? []
            let entries: [ChartDataEntry] = points.compactMap { point in
                guard let x = point["x"] as? Double,
                      let y = point["y"] as? Double else { return nil }
                return ChartDataEntry(x: x, y: y)
            }
            let label = item["label"] as? String ?? ""
            let dataSet = LineChartDataSet(entries: entries, label: label)

            let seriesId = item["id"] as? String
            let styleMap = seriesId.flatMap { perSeriesStyle?[$0] as? [String: Any] } ?? defaultStyle
            applyLineStyle(dataSet: dataSet, styleMap: styleMap)
            dataSets.append(dataSet)
        }

        chartView.data = LineChartData(dataSets: dataSets)

        applyAxis(axis: chartView.xAxis, axisMap: params["xAxis"] as? [String: Any], isXAxis: true)
        applyAxis(axis: chartView.leftAxis, axisMap: params["leftAxis"] as? [String: Any], isXAxis: false)
        applyAxis(axis: chartView.rightAxis, axisMap: params["rightAxis"] as? [String: Any], isXAxis: false)

        applyLegend(legendMap: params["legend"] as? [String: Any])
        applyInteraction(interactionMap: params["interaction"] as? [String: Any])
        applyViewport(viewportMap: params["viewport"] as? [String: Any])
        applyMarker(markerMap: params["marker"] as? [String: Any])
        applyAnimation(animationMap: params["animation"] as? [String: Any])

        chartView.chartDescription.enabled = false
        chartView.notifyDataSetChanged()
    }

    private func applyLineStyle(dataSet: LineChartDataSet, styleMap: [String: Any]?) {
        guard let styleMap else { return }

        if let lineColor = styleMap["lineColor"] as? Int {
            dataSet.colors = [UIColor(argb: lineColor)]
        }
        if let lineWidth = styleMap["lineWidth"] as? Double {
            dataSet.lineWidth = lineWidth
        }
        if let drawCircles = styleMap["drawCircles"] as? Bool {
            dataSet.drawCirclesEnabled = drawCircles
        }
        if let circleColor = styleMap["circleColor"] as? Int {
            dataSet.circleColors = [UIColor(argb: circleColor)]
        } else if let lineColor = styleMap["lineColor"] as? Int {
            dataSet.circleColors = [UIColor(argb: lineColor)]
        }
        if let circleRadius = styleMap["circleRadius"] as? Double {
            dataSet.circleRadius = circleRadius
        }
        if let drawValues = styleMap["drawValues"] as? Bool {
            dataSet.drawValuesEnabled = drawValues
        }
        if let cubic = styleMap["cubic"] as? Bool, cubic == true {
            dataSet.mode = .cubicBezier
        }
        if let lineDashMap = styleMap["lineDash"] as? [String: Any],
           let lengthsRaw = lineDashMap["lengths"] as? [Any],
           lengthsRaw.count >= 2 {
            let lengths = lengthsRaw.compactMap { item -> CGFloat? in
                if let d = item as? Double { return CGFloat(d) }
                if let i = item as? Int { return CGFloat(i) }
                return nil
            }
            if lengths.count >= 2 {
                var phase: CGFloat = 0
                if let p = lineDashMap["phase"] as? Double { phase = CGFloat(p) }
                else if let p = lineDashMap["phase"] as? Int { phase = CGFloat(p) }
                dataSet.lineDashLengths = lengths
                dataSet.lineDashPhase = phase
            }
        }
        let fill = styleMap["fill"] as? String
        if fill == "solid" {
            dataSet.drawFilledEnabled = true
            let fillColor: UIColor
            if let fillColorArgb = styleMap["fillColor"] as? Int {
                fillColor = UIColor(argb: fillColorArgb)
            } else if let lineColorArgb = styleMap["lineColor"] as? Int {
                fillColor = UIColor(argb: lineColorArgb).withAlphaComponent(0.2)
            } else {
                fillColor = (dataSet.colors.first ?? .black).withAlphaComponent(0.2)
            }
            dataSet.fill = ColorFill(color: fillColor)
        } else if fill == "gradient" {
            dataSet.drawFilledEnabled = true
            let lineColorArgb = styleMap["lineColor"] as? Int ?? 0xFF000000
            let topArgb = styleMap["fillColorTop"] as? Int ?? lineColorArgb
            let bottomArgb = styleMap["fillColorBottom"] as? Int ?? 0
            let topColor = UIColor(argb: topArgb)
            let bottomColor = UIColor(argb: bottomArgb)
            let colors = [bottomColor.cgColor, topColor.cgColor]
            let locations: [CGFloat] = [0.0, 1.0]
            guard let gradient = CGGradient(
                colorsSpace: CGColorSpaceCreateDeviceRGB(),
                colors: colors as CFArray,
                locations: locations
            ) else { return }
            dataSet.fill = LinearGradientFill(gradient: gradient, angle: 90)
        }
    }

    private func applyAxis(axis: AxisBase, axisMap: [String: Any]?, isXAxis: Bool) {
        guard let axisMap else { return }
        axis.enabled = axisMap["enabled"] as? Bool ?? true
        axis.drawGridLinesEnabled = axisMap["drawGridLines"] as? Bool ?? true
        axis.drawAxisLineEnabled = axisMap["drawAxisLine"] as? Bool ?? true

        if let min = axisMap["min"] as? Double { axis.axisMinimum = min }
        if let max = axisMap["max"] as? Double { axis.axisMaximum = max }
        if let labelCount = axisMap["labelCount"] as? Int { axis.setLabelCount(labelCount, force: true) }

        if let xAxis = axis as? XAxis, isXAxis {
            xAxis.labelPosition = .bottom
        }
        if let formatter = YAxisValueFormatter(axisMap: axisMap) {
            axis.valueFormatter = formatter
        }
    }

    private func applyLegend(legendMap: [String: Any]?) {
        guard let legendMap else { return }
        let legend = chartView.legend
        legend.enabled = legendMap["enabled"] as? Bool ?? true

        switch legendMap["position"] as? String {
        case "bottom":
            legend.verticalAlignment = .bottom
        case "left":
            legend.verticalAlignment = .top
            legend.horizontalAlignment = .left
        case "right":
            legend.verticalAlignment = .top
            legend.horizontalAlignment = .right
        default:
            legend.verticalAlignment = .top
            legend.horizontalAlignment = .center
        }

        switch legendMap["alignment"] as? String {
        case "start":
            legend.horizontalAlignment = .left
        case "end":
            legend.horizontalAlignment = .right
        default:
            legend.horizontalAlignment = .center
        }

        legend.orientation = .horizontal
        legend.drawInside = false
    }

    private func applyInteraction(interactionMap: [String: Any]?) {
        guard let interactionMap else { return }
        let zoomEnabled = interactionMap["zoomEnabled"] as? Bool ?? true
        let dragEnabled = interactionMap["dragEnabled"] as? Bool ?? true
        let highlightEnabled = interactionMap["highlightEnabled"] as? Bool ?? true

        chartView.scaleXEnabled = zoomEnabled
        chartView.scaleYEnabled = zoomEnabled
        chartView.doubleTapToZoomEnabled = zoomEnabled
        chartView.dragEnabled = dragEnabled
        chartView.highlightPerTapEnabled = highlightEnabled
    }

    private func applyViewport(viewportMap: [String: Any]?) {
        guard let viewportMap else { return }
        if let minRange = viewportMap["visibleXRangeMin"] as? Double {
            chartView.setVisibleXRangeMinimum(minRange)
        }
        if let maxRange = viewportMap["visibleXRangeMax"] as? Double {
            chartView.setVisibleXRangeMaximum(maxRange)
        }
        if let initialX = viewportMap["initialX"] as? Double {
            chartView.moveViewToX(initialX)
        }
        if let offsets = viewportMap["viewPortOffsets"] as? [String: Any] {
            let left = offsets["left"] as? Double ?? 0
            let top = offsets["top"] as? Double ?? 0
            let right = offsets["right"] as? Double ?? 0
            let bottom = offsets["bottom"] as? Double ?? 0
            chartView.setViewPortOffsets(left: left, top: top, right: right, bottom: bottom)
        }
    }

    private func applyMarker(markerMap: [String: Any]?) {
        guard let markerMap else { return }
        let enabled = markerMap["enabled"] as? Bool ?? false
        if !enabled { return }
        let marker = ChartMarkerView()
        marker.chartView = chartView
        chartView.marker = marker
    }

    private func applyAnimation(animationMap: [String: Any]?) {
        guard let animationMap else { return }
        let enabled = animationMap["enabled"] as? Bool ?? false
        if !enabled { return }
        let durationMs = animationMap["durationMs"] as? Double ?? 500
        let duration = durationMs / 1000.0
        switch animationMap["easing"] as? String {
        case "linear":
            chartView.animate(xAxisDuration: duration, easingOption: .linear)
        default:
            chartView.animate(xAxisDuration: duration, easingOption: .easeInOutQuad)
        }
    }
}
